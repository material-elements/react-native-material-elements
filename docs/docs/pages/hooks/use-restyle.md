# useRestyle

### Overview

A utility hook that extracts valid style-related props from a component's props, enabling consistent dynamic styling across custom components.

### Purpose

The useRestyle hook is designed to:

- Filter through a componentâ€™s props.
- Extract style-like props (e.g., `margin`, `padding`, `fontSize`, etc.).
- Return an object that can be passed to the style prop in a React Native component.

It ensures only valid style values (number or string) are included, and skips invalid or non-style-related props.

### Type Signature

```tsx
function useRestyle<T>(props: T & StyledProps): {
  getStyleFromProps: () => Record<string, StyleLike>;
};
```

### How It Works

- `StyledProps`: An interface defining known style-like props (like margin, padding, etc.).
- `isValidStyle`: Internal utility to check if a prop value is a valid style type (`string` or `number`) and not `undefined`, `null`, or a `React element`.
- `getStyleFromProps Function`: Loops through the props and builds a style object from only the valid style props.

### Example Usage

```tsx
const MyComponent = props => {
  const {getStyleFromProps} = useRestyle(props);

  return <View style={getStyleFromProps()} />;
};
```

- If you pass in props like:

```tsx
<MyComponent margin={8} padding={12} customProp="hello" />
```

- Only the valid style props (margin, padding) will be returned by getStyleFromProps():

```tsx
{ margin: 8, padding: 12 }
```

### Return Value

| Property            | Type                      | Description                                                           |
| ------------------- | ------------------------- | --------------------------------------------------------------------- |
| `getStyleFromProps` | Record<string, StyleLike> | Returns an `object` of valid style props (`string` or `number` only). |

### Notes

Ignores props that are:

- `undefined` or `null`
- React elements (e.g., `<Text />`)
- Functions, objects, or other non-style primitives
- Meant to work alongside a StyledProps system or design tokens in a theme-driven environment.

### Related Types

- `StyledProps`: Defines which props are considered `styleable`.
- `StyleLike`: Usually `string` or `number`, representing a single style value.
