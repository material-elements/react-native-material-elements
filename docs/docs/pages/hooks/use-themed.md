# useThemedProps

### Overview

A React hook to resolve dynamic or themed React nodes (e.g., icons or components) based on your design system's current theme colors.

### Purpose

This hook allows developers to:

- Define prop values as either static React elements or dynamic functions that derive from theme colors.
- Automatically resolve those values at render time using the current theme.

It's especially useful for theming UI icons or visual components dynamically depending on the current theme mode (e.g., dark/light).

### Type Signature

```tsx
export type ThemedIconProp = ((theme: Theme) => React.ReactNode) | React.ReactNode;

type ThemedProp<T = any> = T | ((themeColors: ReturnType<typeof useThemeColorsSelector>) => T);

export const useThemedProps = <T extends Record<string, ThemedProp>>(
  props: T,
): { [K in keyof T]: React.ReactNode };
```

### How It Works

- Uses `useThemeColorsSelector Function` to access the current theme’s color tokens.
- Iterates over all props passed into it.
- Resolves each prop:
  - If it's a `function`, it’s called with the theme colors.
  - If it's a `React element`, it’s returned as-is.
  - Otherwise, a `warning` is logged (invalid input).

### Example Usage

```tsx
const MyComponent = ({
  icon,
  activeIcon,
}: {
  icon: ThemedIconProp;
  activeIcon: ThemedIconProp;
}) => {
  const themed = useThemedProps({icon, activeIcon});

  return (
    <View>
      {themed.icon}
      {themed.activeIcon}
    </View>
  );
};
```

- Passing themed props:

```tsx
<MyComponent
  icon={colors => <Icon name="home" color={colors.primary[500]} />}
  activeIcon={<Icon name="home-filled" color="red" />}
/>
```

### Return Value

| Property        | Type                            | Description                                             |
| --------------- | ------------------------------- | ------------------------------------------------------- |
| `resolvedProps` | Record<string, React.ReactNode> | Object with all themed props resolved into React nodes. |

### Warnings & Notes

- Only function and ReactElement types are valid for props.
- If any other type is passed, a console.warn will notify the developer:
  -- icon prop must be either `<Icon />` or `() => <Icon />`. Other values are not valid.
- Use this hook only when dealing with visual components that rely on theming (e.g., icons, badges, labels).

### Related Concepts

- `useThemeColorsSelector`: Hook for accessing current theme’s color palette.
- `Theme`: Your design system’s theme type.
- `ThemedIconProp`: A flexible prop type that accepts either a node or a function of theme.
