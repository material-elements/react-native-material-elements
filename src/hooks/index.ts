export * from './useRestyle';
export * from './useThemedProps';
