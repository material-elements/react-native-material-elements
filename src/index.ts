export * from './components';
export * from './hooks';
export * from './libraries';
export * from './utils';
