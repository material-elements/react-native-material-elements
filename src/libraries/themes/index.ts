export * from './v1';
export * from './V2ThemeContext';
