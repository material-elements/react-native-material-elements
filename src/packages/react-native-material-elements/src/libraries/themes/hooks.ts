import { useContextSelector } from 'use-context-selector';
import { ThemeContext } from './V2ThemeContext';

export const themeErrorMsg = 'Please ensure the ThemeProvider is correctly wrapped around the application.';

export const useThemeFontSelector = () => {
  const themeFonts = useContextSelector(ThemeContext, values => values?.theme?.font);
  if (!themeFonts) {
    throw new Error(`Theme fonts are unavailable. ${themeErrorMsg}`);
  }
  return themeFonts;
};

export const useThemeColorsSelector = () => {
  const themeColors = useContextSelector(ThemeContext, values => values?.theme?.colors);
  if (!themeColors) {
    throw new Error(`Theme colors are unavailable. ${themeErrorMsg}`);
  }
  return themeColors;
};

export const useFontWeightSelector = () => {
  const themeFontWeight = useContextSelector(ThemeContext, values => values?.theme?.fontWeight);
  if (!themeFontWeight) {
    throw new Error(`Theme font weight are unavailable. ${themeErrorMsg}`);
  }
  return themeFontWeight;
};

export const useLatterSpacingSelector = () => {
  const latterSpacing = useContextSelector(ThemeContext, values => values?.theme?.latterSpacing);
  if (!latterSpacing) {
    throw new Error(`Theme latterSpacing are unavailable. ${themeErrorMsg}`);
  }
  return latterSpacing;
};

export const useLineHeightSelector = () => {
  const lineHeight = useContextSelector(ThemeContext, values => values?.theme?.lineHeight);
  if (!lineHeight) {
    throw new Error(`Theme lineHeight are unavailable. ${themeErrorMsg}`);
  }
  return lineHeight;
};

export const useThemeSpacingSelector = () => {
  const themeSpacing = useContextSelector(ThemeContext, values => values?.theme?.spacing);
  if (!themeSpacing) {
    throw new Error(`Theme spacing are unavailable. ${themeErrorMsg}`);
  }
  return themeSpacing;
};

export const useThemeTextConfigSelector = () => {
  return useContextSelector(ThemeContext, values => values?.components?.textProps);
};

export const useThemeBadgeConfigSelector = () => {
  return useContextSelector(ThemeContext, values => values?.components?.badgeProps);
};

export const useThemeButtonConfigSelector = () => {
  return useContextSelector(ThemeContext, values => values?.components?.buttonProps);
};
export const useThemeIconButtonConfigSelector = () => {
  return useContextSelector(ThemeContext, values => values?.components?.iconButtonProps);
};
export const useThemeCheckBoxConfigSelector = () => {
  return useContextSelector(ThemeContext, values => values?.components?.checkBoxProps);
};

export const useThemeChipConfigSelector = () => {
  return useContextSelector(ThemeContext, values => values?.components?.chipProps);
};

export const useThemeDividerConfigSelector = () => {
  return useContextSelector(ThemeContext, values => values?.components?.dividerProps);
};

export const useThemeListConfigSelector = () => {
  return useContextSelector(ThemeContext, values => values?.components?.listProps);
};

export const useThemeListItemConfigSelector = () => {
  return useContextSelector(ThemeContext, values => values?.components?.listItemProps);
};

export const useThemePaginationConfigSelector = () => {
  return useContextSelector(ThemeContext, values => values?.components?.paginationProps);
};

export const useThemeRadioConfigSelector = () => {
  return useContextSelector(ThemeContext, values => values?.components?.radioProps);
};

export const useThemeSwitchConfigSelector = () => {
  return useContextSelector(ThemeContext, values => values?.components?.switchProps);
};

export const useThemeTextFieldConfigSelector = () => {
  return useContextSelector(ThemeContext, values => values?.components?.textFieldProps);
};

export const useThemeIconInputConfigSelector = () => {
  return useContextSelector(ThemeContext, values => values?.components?.iconInputProps);
};
