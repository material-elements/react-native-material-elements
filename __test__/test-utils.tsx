import React from 'react';
import { render, RenderOptions } from '@testing-library/react-native';
import { ThemeProvider } from '../src';

export const ThemeWrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  return <ThemeProvider>{children}</ThemeProvider>;
};

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: ThemeWrapper, ...options });

export * from '@testing-library/react-native';
export { customRender as render };
