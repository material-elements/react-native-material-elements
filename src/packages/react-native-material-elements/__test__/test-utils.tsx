import React from 'react';
import { render, RenderOptions } from '@testing-library/react-native';
import { ThemeProvider } from '../src';
import { ThemeProviderProps } from '../src/libraries/types';

interface CustomRenderOptions extends Omit<RenderOptions, 'wrapper'> {
  themeProps?: Partial<ThemeProviderProps>;
}

export const ThemeWrapper: React.FC<ThemeProviderProps> = ({ children, ...props }) => {
  return <ThemeProvider {...props}>{children}</ThemeProvider>;
};

export const customRender = (ui: React.ReactElement, options?: CustomRenderOptions) => {
  const { themeProps, ...rest } = options || {};
  return render(ui, {
    wrapper: wrapperProps => <ThemeWrapper {...themeProps}>{wrapperProps.children}</ThemeWrapper>,
    ...rest,
  });
};

export * from '@testing-library/react-native';
export { customRender as render };
