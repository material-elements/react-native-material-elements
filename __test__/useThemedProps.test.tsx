import { renderHook } from '@testing-library/react-hooks';
import React, { isValidElement } from 'react';
import { Text, View } from 'react-native';
import { useThemedProps } from '../src/hooks';
import { render, ThemeWrapper } from './test-utils';
import { Theme } from '../src/libraries/themes/types';
import { green } from '../src';

describe('useThemedProps', () => {
  it('should match icon component snapshot correctly', () => {
    const props = {
      icon: () => <View />,
    };
    const { result } = renderHook(() => useThemedProps(props), { wrapper: ThemeWrapper });
    const { toJSON } = render(result.current.icon);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render icon component correctly', () => {
    const props = {
      icon: () => <View />,
    };
    const { result } = renderHook(() => useThemedProps(props), { wrapper: ThemeWrapper });

    const renderedIcon = result.current.icon;
    expect(isValidElement(renderedIcon)).toBe(true);
    expect(renderedIcon).toBeDefined();
  });

  it('should not apply theme-based style override', () => {
    const props = {
      style: { color: 'red' },
    };

    const { result } = renderHook(() => useThemedProps(props), { wrapper: ThemeWrapper });

    expect(result.current.style).toBeUndefined();
  });

  it('should handle no icon provided gracefully', () => {
    const props = {};
    const { result } = renderHook(() => useThemedProps(props), { wrapper: ThemeWrapper });

    expect(result.current.icon).toBeUndefined();
  });

  it('should support label as a ReactNode', () => {
    const props = {
      label: <Text>Label</Text>,
    };

    const { result } = renderHook(() => useThemedProps(props), { wrapper: ThemeWrapper });

    expect(isValidElement(result.current.label)).toBe(true);
  });

  it('should preserve custom props', () => {
    const props = {
      customKey: 'customValue',
    };

    const { result } = renderHook(() => useThemedProps(props), { wrapper: ThemeWrapper });

    expect(result.current.customKey).toBeUndefined();
  });

  it('should memoize icon function result', () => {
    const iconFn = jest.fn(() => <View />);
    const props = { icon: iconFn };

    const { rerender } = renderHook(() => useThemedProps(props), { wrapper: ThemeWrapper });

    expect(iconFn).toHaveBeenCalledTimes(1);
    rerender();
    expect(iconFn).toHaveBeenCalledTimes(1);
  });

  it('should resolve function props using themeColors', () => {
    const props = {
      icon: (theme: Theme) => <View style={{ backgroundColor: theme.green[500] }} />,
    };

    const { result } = renderHook(() => useThemedProps(props), { wrapper: ThemeWrapper });

    const iconComponent = result.current.icon;

    expect(iconComponent).toBeDefined();
    expect(iconComponent.props.style.backgroundColor).toEqual(green[500]);
  });

  it('should render component correctly if user passed component directly', () => {
    const props = {
      icon: <View />,
    };
    const { result } = renderHook(() => useThemedProps(props), { wrapper: ThemeWrapper });

    const iconComponent = result.current.icon;
    expect(iconComponent).toBeDefined();
    const { toJSON } = render(iconComponent);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should return an empty object when props are empty', () => {
    const { result } = renderHook(() => useThemedProps({}), { wrapper: ThemeWrapper });

    expect(result.current).toEqual({});
  });

  it('should return valid React elements when ReactNode props are provided', () => {
    const StaticIcon = <View testID="static-icon" />;
    const props = {
      icon: StaticIcon,
      label: <View testID="label" />,
    };

    const { result } = renderHook(() => useThemedProps(props), { wrapper: ThemeWrapper });

    expect(result.current.icon).toBe(StaticIcon);
    expect(result.current.label).toBeDefined();
  });

  it('should support themed prop returning a primitive value', () => {
    const props = {
      fontSize: () => 16,
    };

    const { result } = renderHook(() => useThemedProps(props), { wrapper: ThemeWrapper });

    expect(result.current.fontSize).toBe(16);
  });

  it('should evaluate icon function only once even across re-renders', () => {
    const iconFn = jest.fn(() => <View />);
    const props = { icon: iconFn };

    const { rerender } = renderHook(() => useThemedProps(props), {
      wrapper: ThemeWrapper,
      initialProps: props,
    });

    rerender(props);
    rerender(props);

    expect(iconFn).toHaveBeenCalledTimes(1);
  });

  it('should support themed props returning null', () => {
    const props = {
      icon: () => null,
    };

    const { result } = renderHook(() => useThemedProps(props), { wrapper: ThemeWrapper });

    expect(result.current.icon).toBeNull();
  });

  it('should set undefined if function does not return a value', () => {
    const props = {
      icon: () => {},
    };

    const { result } = renderHook(() => useThemedProps(props), { wrapper: ThemeWrapper });

    expect(result.current.icon).toBeUndefined();
  });

  it('should ignore deep nested functions (not supported by current impl)', () => {
    const props = {
      styles: {
        color: (theme: Theme) => theme.primary,
      },
    };

    const { result } = renderHook(() => useThemedProps(props as any), { wrapper: ThemeWrapper });

    expect(result.current.styles).toBeUndefined();
  });

  it('should not return array props as-is', () => {
    const props = {
      values: [1, 2, 3],
    };

    const { result } = renderHook(() => useThemedProps(props), { wrapper: ThemeWrapper });

    expect(result.current.values).toBeUndefined();
  });

  it('should warn if icon is a primitive value', () => {
    const warnSpy = jest.spyOn(console, 'warn').mockImplementation(() => {});

    const props = {
      icon: 'not-a-component' as any,
    };

    renderHook(() => useThemedProps(props), { wrapper: ThemeWrapper });

    expect(warnSpy).toHaveBeenCalledWith('icon prop must be either <Icon /> or () => <Icon />. Other values are not valid.');

    warnSpy.mockRestore();
  });

  it('should warn if icon is a plain object', () => {
    const warnSpy = jest.spyOn(console, 'warn').mockImplementation(() => {});

    const props = {
      icon: { something: true } as any,
    };

    renderHook(() => useThemedProps(props), { wrapper: ThemeWrapper });

    expect(warnSpy).toHaveBeenCalledWith('icon prop must be either <Icon /> or () => <Icon />. Other values are not valid.');

    warnSpy.mockRestore();
  });

  it('should not warn if icon is a valid React element', () => {
    const warnSpy = jest.spyOn(console, 'warn').mockImplementation(() => {});

    const props = {
      icon: () => <View />,
    };

    renderHook(() => useThemedProps(props));
    expect(warnSpy).not.toHaveBeenCalled();
    warnSpy.mockRestore();
  });
});
