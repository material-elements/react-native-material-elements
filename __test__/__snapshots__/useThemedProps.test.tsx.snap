// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`useThemedProps should match icon component snapshot correctly 1`] = `<View />`;

exports[`useThemedProps should render component correctly if user passed component directly 1`] = `<View />`;
